In [9]:
# Code snippet below is from Catherine Magsino
import sqlite3
import pandas        as pd
import pandas.io.sql as pd_sql

# Connect to the sql table
conn   = sqlite3.connect("./dataset_soccer/database.sqlite")
cursor = conn.cursor()
    
#query = "SELECT * FROM ;"

In [10]:
# This tells you what tables are available in the SQL database file
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
print(cursor.fetchall())

[('sqlite_sequence',), ('Player_Attributes',), ('Player',), ('Match',), ('League',), ('Country',), ('Team',), ('Team_Attributes',)]


In [11]:
# Function to print out all table columns
def table_columns(db, table_name):
    curs = db.cursor()
    sql = "select * from %s where 1=0;" % table_name
    curs.execute(sql)
    return [d[0] for d in curs.description]

In [15]:
#table_columns(conn, 'Player_Attributes')

In [61]:
pd.read_sql_query('SELECT * FROM Match',conn)

id  country_id  league_id     season  stage                 date  \
0          1           1          1  2008/2009      1  2008-08-17 00:00:00   
1          2           1          1  2008/2009      1  2008-08-16 00:00:00   
2          3           1          1  2008/2009      1  2008-08-16 00:00:00   
3          4           1          1  2008/2009      1  2008-08-17 00:00:00   
4          5           1          1  2008/2009      1  2008-08-16 00:00:00   
5          6           1          1  2008/2009      1  2008-09-24 00:00:00   
6          7           1          1  2008/2009      1  2008-08-16 00:00:00   
7          8           1          1  2008/2009      1  2008-08-16 00:00:00   
8          9           1          1  2008/2009      1  2008-08-16 00:00:00   
9         10           1          1  2008/2009     10  2008-11-01 00:00:00   
10        11           1          1  2008/2009     10  2008-10-31 00:00:00   
11        12           1          1  2008/2009     10  2008-11-02 00:00:00   
12        13           1          1  2008/2009     10  2008-11-01 00:00:00   
13        14           1          1  2008/2009     10  2008-11-01 00:00:00   
14        15           1          1  2008/2009     10  2008-11-01 00:00:00   
15        16           1          1  2008/2009     10  2008-11-01 00:00:00   
16        17           1          1  2008/2009     10  2008-11-01 00:00:00   
17        18           1          1  2008/2009     10  2008-11-02 00:00:00   
18        19           1          1  2008/2009     11  2008-11-08 00:00:00   
19        20           1          1  2008/2009     11  2008-11-08 00:00:00   
20        21           1          1  2008/2009     11  2008-11-09 00:00:00   
21        22           1          1  2008/2009     11  2008-11-07 00:00:00   
22        23           1          1  2008/2009     11  2008-11-08 00:00:00   
23        24           1          1  2008/2009     11  2008-11-08 00:00:00   
24        25           1          1  2008/2009     11  2008-11-08 00:00:00   
25        26           1          1  2008/2009     11  2008-11-08 00:00:00   
26        27           1          1  2008/2009     11  2008-11-09 00:00:00   
27        28           1          1  2008/2009     12  2008-11-16 00:00:00   
28        29           1          1  2008/2009     12  2008-11-15 00:00:00   
29        30           1          1  2008/2009     12  2008-11-15 00:00:00   
...      ...         ...        ...        ...    ...                  ...   
25949  25950       24558      24558  2015/2016      4  2015-08-08 00:00:00   
25950  25951       24558      24558  2015/2016      4  2015-08-08 00:00:00   
25951  25952       24558      24558  2015/2016      4  2015-08-09 00:00:00   
25952  25953       24558      24558  2015/2016      4  2015-08-09 00:00:00   
25953  25954       24558      24558  2015/2016      4  2015-08-09 00:00:00   
25954  25955       24558      24558  2015/2016      5  2015-08-12 00:00:00   
25955  25956       24558      24558  2015/2016      5  2015-08-12 00:00:00   
25956  25957       24558      24558  2015/2016      5  2015-08-12 00:00:00   
25957  25958       24558      24558  2015/2016      5  2015-08-13 00:00:00   
25958  25959       24558      24558  2015/2016      5  2015-08-13 00:00:00   
25959  25960       24558      24558  2015/2016      6  2015-08-22 00:00:00   
25960  25961       24558      24558  2015/2016      6  2015-08-23 00:00:00   
25961  25962       24558      24558  2015/2016      6  2015-08-23 00:00:00   
25962  25963       24558      24558  2015/2016      6  2015-08-23 00:00:00   
25963  25964       24558      24558  2015/2016      6  2015-08-22 00:00:00   
25964  25965       24558      24558  2015/2016      7  2015-08-29 00:00:00   
25965  25966       24558      24558  2015/2016      7  2015-08-29 00:00:00   
25966  25967       24558      24558  2015/2016      7  2015-08-30 00:00:00   
25967  25968       24558      24558  2015/2016      7  2015-08-30 00:00:00   
25968  25969       24558      24

In [18]:
table_columns(conn, 'Match')

['id',
 'country_id',
 'league_id',
 'season',
 'stage',
 'date',
 'match_api_id',
 'home_team_api_id',
 'away_team_api_id',
 'home_team_goal',
 'away_team_goal',
 'home_player_X1',
 'home_player_X2',
 'home_player_X3',
 'home_player_X4',
 'home_player_X5',
 'home_player_X6',
 'home_player_X7',
 'home_player_X8',
 'home_player_X9',
 'home_player_X10',
 'home_player_X11',
 'away_player_X1',
 'away_player_X2',
 'away_player_X3',
 'away_player_X4',
 'away_player_X5',
 'away_player_X6',
 'away_player_X7',
 'away_player_X8',
 'away_player_X9',
 'away_player_X10',
 'away_player_X11',
 'home_player_Y1',
 'home_player_Y2',
 'home_player_Y3',
 'home_player_Y4',
 'home_player_Y5',
 'home_player_Y6',
 'home_player_Y7',
 'home_player_Y8',
 'home_player_Y9',
 'home_player_Y10',
 'home_player_Y11',
 'away_player_Y1',
 'away_player_Y2',
 'away_player_Y3',
 'away_player_Y4',
 'away_player_Y5',
 'away_player_Y6',
 'away_player_Y7',
 'away_player_Y8',
 'away_player_Y9',
 'away_player_Y10',
 'away_player

In [52]:
table_columns(conn, 'Team_Attributes')

['id',
 'team_fifa_api_id',
 'team_api_id',
 'date',
 'buildUpPlaySpeed',
 'buildUpPlaySpeedClass',
 'buildUpPlayDribbling',
 'buildUpPlayDribblingClass',
 'buildUpPlayPassing',
 'buildUpPlayPassingClass',
 'buildUpPlayPositioningClass',
 'chanceCreationPassing',
 'chanceCreationPassingClass',
 'chanceCreationCrossing',
 'chanceCreationCrossingClass',
 'chanceCreationShooting',
 'chanceCreationShootingClass',
 'chanceCreationPositioningClass',
 'defencePressure',
 'defencePressureClass',
 'defenceAggression',
 'defenceAggressionClass',
 'defenceTeamWidth',
 'defenceTeamWidthClass',
 'defenceDefenderLineClass']

In [22]:
pd.read_sql_query('SELECT * FROM Team LIMIT 5',conn)

id  team_api_id  team_fifa_api_id     team_long_name team_short_name
0   1         9987               673           KRC Genk             GEN
1   2         9993               675       Beerschot AC             BAC
2   3        10000             15005   SV Zulte-Waregem             ZUL
3   4         9994              2007   Sporting Lokeren             LOK
4   5         9984              1750  KSV Cercle Brugge             CEB

In [20]:
pd.read_sql_query('SELECT * FROM League',conn)

id  country_id                      name
0       1           1    Belgium Jupiler League
1    1729        1729    England Premier League
2    4769        4769            France Ligue 1
3    7809        7809     Germany 1. Bundesliga
4   10257       10257             Italy Serie A
5   13274       13274    Netherlands Eredivisie
6   15722       15722        Poland Ekstraklasa
7   17642       17642  Portugal Liga ZON Sagres
8   19694       19694   Scotland Premier League
9   21518       21518           Spain LIGA BBVA
10  24558       24558  Switzerland Super League

### Q1: Which team scored the most points when playing at home? 

In [50]:
# Attempt Question 1 here
#Which team scored the most points when playing at home?

sql_text_query = '''
SELECT Team.team_long_name, SUM(Match.home_team_goal)
FROM [Team] 
JOIN [Match] ON Team.team_api_id=Match.home_team_api_id
GROUP BY Team.team_long_name
ORDER BY SUM(Match.home_team_goal) DESC LIMIT 5;
'''

pd.read_sql_query(sql_text_query,conn)

#'SELECT Team.team_long_name, \ 
#Match.home_team_goal FROM Team JOIN Match ON Team.id=Match.id;'

team_long_name  SUM(Match.home_team_goal)
0    Real Madrid CF                        505
1      FC Barcelona                        495
2            Celtic                        389
3  FC Bayern Munich                        382
4               PSV                        370

Real Madrid CF with 505 home goals

### Q2: Did this team also score the most points when playing away?

In [51]:
# Question 2
sql_text_query = '''
SELECT Team.team_long_name, SUM(Match.away_team_goal)
FROM [Team] 
JOIN [Match] ON Team.team_api_id=Match.away_team_api_id
GROUP BY Team.team_long_name
ORDER BY SUM(Match.away_team_goal) DESC LIMIT 5;
'''

pd.read_sql_query(sql_text_query,conn)

team_long_name  SUM(Match.away_team_goal)
0    FC Barcelona                        354
1  Real Madrid CF                        338
2          Celtic                        306
3            Ajax                        287
4             PSV                        282

No, FC Barcelona scored the most away goals with 354 goals

### Q3: How many matches resulted in a tie?

In [56]:
pd.read_sql_query('SELECT match_api_id, home_team_api_id,away_team_api_id, home_team_goal, away_team_goal FROM Match LIMIT 5',conn)

match_api_id  home_team_api_id  away_team_api_id  home_team_goal  \
0        492473              9987              9993               1   
1        492474             10000              9994               0   
2        492475              9984              8635               0   
3        492476              9991              9998               5   
4        492477              7947              9985               1   

   away_team_goal  
0               1  
1               0  
2               3  
3               0  
4               3

In [104]:
sql_text_query = '''
SELECT COUNT(home_team_goal=away_team_goal) AS X
FROM [Match]
WHERE home_team_goal=away_team_goal

'''

pd.read_sql_query(sql_text_query,conn)

X
0  6596

### Q4: How many players have Smith for their last name? How many have 'smith' anywhere in their name?

In [70]:
sql_text_query = '''
SELECT id, COUNT(player_name)
FROM [Player]
WHERE player_name LIKE '% Smith'

'''
#WHERE home_team_goal=away_team_goal


pd.read_sql_query(sql_text_query,conn)

id  COUNT(player_name)
0  129                  15

In [73]:
sql_text_query = '''
SELECT id, COUNT(player_name)
FROM [Player]
WHERE player_name LIKE '%Smith%'

'''
pd.read_sql_query(sql_text_query,conn)

id  COUNT(player_name)
0  129                  18

### Q5: What was the median tie score? Use the value determined in the previous question for the number of tie games. Hint: PostgreSQL does not have a median function. Instead, think about the steps required to calculate a median and use the WITH command to store stepwise results as a table and then operate on these results.


In [137]:
sql_text_query = '''
CREATE TABLE [NewTable3] AS SELECT match_api_id, home_team_goal=away_team_goal AS Tie
FROM [Match] WHERE home_team_goal=away_team_goal;

'''

pd.read_sql_query(sql_text_query,conn)

TypeError: 'NoneType' object is not iterable

In [139]:
# From Stackoverflow --- https://stackoverflow.com/questions/15763965/how-can-i-calculate-the-median-of-values-in-sqlite

#SELECT AVG(x)
#FROM (SELECT x
#      FROM MyTable
#      ORDER BY x
#      LIMIT 2
#      OFFSET (SELECT (COUNT(*) - 1) / 2
#              FROM MyTable))


sql_text_query = '''

SELECT ROUND(AVG(home_team_goal), 2) AS median_item_count  
FROM (  SELECT home_team_goal
        FROM ( SELECT home_team_goal,
           COUNT(*) OVER() AS row_count,
           ROW_NUMBER() OVER(ORDER BY home_team_goal=away_team_goal) AS row_number
    FROM [Match]
    WHERE home_team_goal=away_team_goal <> 0
  ) x
WHERE row_number IN ((row_count + 1)/2, (row_count + 2)/2)  
) y
;
'''
pd.read_sql_query(sql_text_query,conn)

#OFFSET (SELECT (COUNT(*) - 1) / 2 FROM [Match])

median_item_count
0                1.0

### Q6: What percentage of players prefer their left or right foot? Hint: Calculate either the right or left foot, whichever is easier based on how you setup the problem.

In [93]:
sql_text_query = '''
SELECT *
FROM [Player_Attributes]


'''
pd.read_sql_query(sql_text_query,conn)

id  player_fifa_api_id  player_api_id                 date  \
0            1              218353         505942  2016-02-18 00:00:00   
1            2              218353         505942  2015-11-19 00:00:00   
2            3              218353         505942  2015-09-21 00:00:00   
3            4              218353         505942  2015-03-20 00:00:00   
4            5              218353         505942  2007-02-22 00:00:00   
5            6              189615         155782  2016-04-21 00:00:00   
6            7              189615         155782  2016-04-07 00:00:00   
7            8              189615         155782  2016-01-07 00:00:00   
8            9              189615         155782  2015-12-24 00:00:00   
9           10              189615         155782  2015-12-17 00:00:00   
10          11              189615         155782  2015-10-16 00:00:00   
11          12              189615         155782  2015-09-25 00:00:00   
12          13              189615         155782  2015-09-21 00:00:00   
13          14              189615         155782  2015-01-09 00:00:00   
14          15              189615         155782  2014-12-05 00:00:00   
15          16              189615         155782  2014-11-07 00:00:00   
16          17              189615         155782  2014-09-18 00:00:00   
17          18              189615         155782  2014-05-02 00:00:00   
18          19              189615         155782  2014-04-04 00:00:00   
19          20              189615         155782  2014-03-14 00:00:00   
20          21              189615         155782  2013-12-13 00:00:00   
21          22              189615         155782  2013-11-08 00:00:00   
22          23              189615         155782  2013-10-04 00:00:00   
23          24              189615         155782  2013-09-20 00:00:00   
24          25              189615         155782  2013-05-03 00:00:00   
25          26              189615         155782  2013-03-22 00:00:00   
26          27              189615         155782  2013-03-15 00:00:00   
27          28              189615         155782  2013-02-22 00:00:00   
28          29              189615         155782  2013-02-15 00:00:00   
29          30              189615         155782  2012-08-31 00:00:00   
...        ...                 ...            ...                  ...   
183948  183949              164680         111182  2012-02-22 00:00:00   
183949  183950              164680         111182  2011-08-30 00:00:00   
183950  183951              164680         111182  2011-02-22 00:00:00   
183951  183952              164680         111182  2010-08-30 00:00:00   
183952  183953              164680         111182  2007-02-22 00:00:00   
183953  183954              111191          36491  2011-02-22 00:00:00   
183954  183955              111191          36491  2010-08-30 00:00:00   
183955  183956              111191          36491  2009-08-30 00:00:00   
183956  183957              111191          36491  2009-02-22 00:00:00   
183957  183958              111191          36491  2008-08-30 00:00:00   
183958  183959              111191          36491  2007-08-30 00:00:00   
183959  183960              111191          36491  2007-02-22 00:00:00   
183960  183961               47058          35506  2011-02-22 00:00:00   
183961  183962               47058          35506  2010-08-30 00:00:00   
183962  183963               47058          35506  2010-02-22 00:00:00   
183963  183964               47058          35506  2009-08-30 00:00:00   
183964  183965               47058          35506  2009-02-22 00:00:00   
183965  183966               47058          35506  2008-08-30 00:00:00   
183966  183967               47058          35506  2007-08-30 00:00:00   
183967  183968               47058          35506  2007-02-22 00:00:00   
183968  183969              102359          39902  2012-08-31 00:00:00   
183969  183970              102359          39902  2012-02-22 00:00:00   
183970  183971  

In [140]:
sql_text_query = '''
SELECT preferred_foot, (Count(preferred_foot)*100 / (Select Count(*) From [Player_Attributes]))
FROM [Player_Attributes]
WHERE preferred_foot='right';
'''
pd.read_sql_query(sql_text_query,conn)

#Select Grade, (Count(Grade)* 100 / (Select Count(*) From MyTable)) as Score
#From MyTable
#Group By Grade

preferred_foot  (Count(preferred_foot)*100 / (Select Count(*) From
0          right                                                 75